commons-math/src/main/java/org/apache/commons/math4/optim/nonlinear/scalar/noderiv/BOBYQAOptimizer.java

```diff
@@ -0,0 +1,2475 @@
+
+    /**
+     *     A version of the truncated conjugate gradient is applied. If a line
+     *     search is restricted by a constraint, then the procedure is restarted,
+     *     the values of the variables that are at their bounds being fixed. If
+     *     the trust region boundary is reached, then further changes may be made
+     *     to D, each one being in the two dimensional space that is spanned
+     *     by the current D and the gradient of Q at XOPT+D, staying on the trust
+     *     region boundary. Termination occurs when the reduction in Q seems to
+     *     be close to the greatest reduction that can be achieved.
+     *     The arguments N, NPT, XPT, XOPT, GOPT, HQ, PQ, SL and SU have the same
+     *       meanings as the corresponding arguments of BOBYQB.
+     *     DELTA is the trust region radius for the present calculation, which
+     *       seeks a small value of the quadratic model within distance DELTA of
+     *       XOPT subject to the bounds on the variables.
+     *     XNEW will be set to a new vector of variables that is approximately
+     *       the one that minimizes the quadratic model within the trust region
+     *       subject to the SL and SU constraints on the variables. It satisfies
+     *       as equations the bounds that become active during the calculation.
+     *     D is the calculated trial step from XOPT, generated iteratively from an
+     *       initial value of zero. Thus XNEW is XOPT+D after the final iteration.
+     *     GNEW holds the gradient of the quadratic model at XOPT+D. It is updated
+     *       when D is updated.
+     *     xbdi.get( is a working space vector. For I=1,2,...,N, the element xbdi.get((I) is
+     *       set to -1.0, 0.0, or 1.0, the value being nonzero if and only if the
+     *       I-th variable has become fixed at a bound, the bound being SL(I) or
+     *       SU(I) in the case xbdi.get((I)=-1.0 or xbdi.get((I)=1.0, respectively. This
+     *       information is accumulated during the construction of XNEW.
+     *     The arrays S, HS and HRED are also used for working space. They hold the
+     *       current search direction, and the changes in the gradient of Q along S
+     *       and the reduced D, respectively, where the reduced D is the same as D,
+     *       except that the components of the fixed variables are zero.
+     *     DSQ will be set to the square of the length of XNEW-XOPT.
+     *     CRVMIN is set to zero if D reaches the trust region boundary. Otherwise
+     *       it is set to the least curvature of H that occurs in the conjugate
+     *       gradient searches that are not restricted by any constraints. The
+     *       value CRVMIN=-1.0D0 is set, however, if all of these searches are
+     *       constrained.
+     * @param delta
+     * @param gnew
+     * @param xbdi
+     * @param s
+     * @param hs
+     * @param hred
+     */
+    private double[] trsbox(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 75. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.